Lib/test/test_hashlib.py

```diff
@@ -274,7 +274,10 @@ def test_clinic_signature(self):
                 with self.assertWarnsRegex(DeprecationWarning,
                                            DEPRECATED_STRING_PARAMETER):
                     hashlib.new(digest_name, string=b'')
-                if self._hashlib:
+                # when using a combination of libcrypto and interned hash
+                # implementations, we need to make sure that _hashlib contains
+                # the constructor we're testing
```

Comment:
```suggestion
                # Make sure that _hashlib contains the constructor
                # to test when using a combination of libcrypto and
                # interned hash implementations.
```

By "interned" hash implementations, do you mean HACL* implementations or others? because `hash_constructors` should actually only contain the the constructors that are "correct"?

Also, I think you can use `_hashlib.algorithms_available` instead of `_constructors`, but I don't remember well now. I'm leaving so I won't be able to review more.

Reply:
> By "interned" hash implementations, do you mean HACL* implementations or others?

I'm not sure which are HACL and which are not, but by "interned" I mean all the hash functions that use source in CPython's tree for hash implementations instead of a linking to libcrypto.

> because `hash_constructors` should actually only contain the the constructors that are "correct"

In PR #135402, I observed `hash_constructors` containing `blake2b`/`blake2s` constructors while `_hashlib` did not, causing failures when this test tried to invoke it.

> Also, I think you can use `_hashlib.algorithms_available` instead of `_constructors`

Hm, I don't see an `algorithms_available` attr of `_hashlib`:

```
$ ./python
Python 3.15.0a0 (heads/add-awslc-ci-job-dirty:be1b72c13b2, Jun 12 2025, 15:30:29) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ssl
>>> ssl.OPENSSL_VERSION
'AWS-LC 1.52.1'
>>> import _hashlib
>>> dir(_hashlib)
['HASH', 'HASHXOF', 'HMAC', 'UnsupportedDigestmodError', '_GIL_MINSIZE', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '_constructors', 'compare_digest', 'get_fips_mode', 'hmac_digest', 'hmac_new', 'new', 'openssl_md5', 'openssl_md_meth_names', 'openssl_sha1', 'openssl_sha224', 'openssl_sha256', 'openssl_sha384', 'openssl_sha3_224', 'openssl_sha3_256', 'openssl_sha3_384', 'openssl_sha3_512', 'openssl_sha512', 'openssl_shake_128', 'openssl_shake_256', 'pbkdf2_hmac', 'scrypt']
>>> _hashlib.algorithms_available
Traceback (most recent call last):
  File "<python-input-2>", line 1, in <module>
    _hashlib.algorithms_available
AttributeError: module '_hashlib' has no attribute 'algorithms_available'
```

Reply:
Sorry, it was `hashlib.algorithms_available` but it wouldn't help. I *think* you can use `h = constructor()` and check `h.name in self.supported_hash_names` 

Can you also modify `test_clinic_signature_errors` please?

Reply:
>  I think you can use` h = constructor()` and check `h.name in self.supported_hash_names`

I don't think that works because `supported_hash_names` [is constant](https://github.com/python/cpython/blob/61532b4bc783c1570c0fd7fb2ff5f8157cf64e82/Lib/test/test_hashlib.py#L110) and independent of what was linked into `_hashlib`. We need to check if the constructor is actually present in `_hashlib`.

> Can you also modify `test_clinic_signature_errors` please?

No problem, done in 27ca72f.